Tests/ApolloCodegenTests/Configuration/ApolloCodegenConfiguration+SchemaCustomizationTests.swift

```diff
@@ -0,0 +1,605 @@
+    testJSONEncoder = nil
+    try super.tearDownWithError()
+  }
+
+  // MARK: - Custom Type Names
+  
+  func test__encodeType_withCustomName() throws {
+    // given
+    let subject = ApolloCodegenConfiguration.SchemaCustomization(
+      customTypeNames: [
+        "MyObject": .type(name: "CustomObject")
+      ]
+    )
+    
+    let expected = """
+    {
+      "customTypeNames" : {
+        "MyObject" : "CustomObject"
+      }
+    }
+    """
+    
+    // when
+    let encodedJSON = try testJSONEncoder.encode(subject)
+    let actual = encodedJSON.asString
+    
+    //then
+    expect(actual).to(equalLineByLine(expected))
+  }
+  
+  func test__decodeType_withCustomName() throws {
+    // given
+    let subject = """
+    {
+      "customTypeNames" : {
+        "MyObject" : "CustomObject"
+      }
+    }
+    """
+    
+    let expected = ApolloCodegenConfiguration.SchemaCustomization(
+      customTypeNames: [
+        "MyObject": .type(name: "CustomObject")
+      ]
+    )
+    
+    // when
+    let actual = try JSONDecoder().decode(ApolloCodegenConfiguration.SchemaCustomization.self, from: subject.asData)
+    
+    //then
+    expect(actual).to(equal(expected))
+  }
+  
+  func test__encodeEmptyType_shouldThrowError() throws {
+    // given
+    let subject = ApolloCodegenConfiguration.SchemaCustomization(
+      customTypeNames: [
+        "MyObject": .type(name: "")
+      ]
+    )
+    
+    //then
+    expect {
+      _ = try self.testJSONEncoder.encode(subject)
+    }.to(throwError { error in
+      guard case let ApolloCodegenConfiguration.SchemaCustomization.Error.emptyCustomization(type) = error else {
+        fail("Expected .emptyCustomization, got .\(error)")
+        return
+      }
+      expect(type).to(equal("MyObject"))
+    })
+  }
+  
+  func test__decodeEmptyType_shouldThrowError() throws {
+    // given
+    let subject = """
+    {
+      "customTypeNames" : {
+        "MyObject" : ""
+      }
+    }
+    """
+    
+    ///then
+    expect {
+      try JSONDecoder().decode(ApolloCodegenConfiguration.SchemaCustomization.self, from: subject.asData)
+    }.to(throwError { error in
+      guard case let ApolloCodegenConfiguration.SchemaCustomization.Error.emptyCustomization(type) = error else {
+        fail("Expected .emptyCustomization, got .\(error)")
+        return
+      }
+      expect(type).to(equal("MyObject"))
+    })
+  }
+  
+  // MARK: - Custom Enum Names
+  
+  func test__encodeEnum_withCustomNameAndCases() throws {
+    // given
+    let subject = ApolloCodegenConfiguration.SchemaCustomization(
+      customTypeNames: [
+        "MyEnum": .enum(
+          name: "CustomEnum",
+          cases: [
+            "CaseOne": "CustomCaseOne"
+          ]
+        )
+      ]
+    )
+    
+    let expected = """
+    {
+      "customTypeNames" : {
+        "MyEnum" : {
+          "enum" : {
+            "cases" : {
+              "CaseOne" : "CustomCaseOne"
+            },
+            "name" : "CustomEnum"
```

Comment:
Nit: would be nice is `name` came before `cases`. I know the encoding is alphabetical by default, but we could manually change this. Not that important though.

Reply:
Would have to test to check but I think in actual use that may be the case since we encode name first in the implementation, but in the Tests specifically we have the encoder setup to do alphabetical order.